File: 2016/python/07.py
# Advent of Code 2016, day 7
# (c) aichingert

def contains_abba(s):
    for i in range(len(s)-3):
        if s[i] == s[i+1]:
            continue
        if s[i] == s[i+3] and s[i+1] == s[i+2]:
            return True
    return False

def part1(inp):
    ans = 0
    for l in inp:
        left,rest = l.strip().split('[',1)
        v = contains_abba(left)

        for sp in rest.split('['):
            h, a = sp.split(']',1)
            if contains_abba(h):
                v = False
                break
            elif contains_abba(a):
                v = True
        if v:
            ans += 1

    return ans

def contains_ssl(addr, hypr):
    for i in range(len(addr)):
        for j in range(len(addr[i])-2):
            if addr[i][j] == addr[i][j+1]:
                continue
            if addr[i][j] == addr[i][j+2]:
                for hyp in hypr:
                    for k in range(len(hyp)-2):
                        if hyp[k] == hyp[k+2] and hyp[k] == addr[i][j+1] and hyp[k+1] == addr[i][j]:
                            return True
    return False

def part2(inp):
    ans = 0
    for l in inp:
        left,rest = l.strip().split('[',1)
        addr = [left]
        hypr = []
        for sp in rest.split('['):
            h,a = sp.split(']',1)
            addr.append(a)
            hypr.append(h)
        if contains_ssl(addr, hypr):
            ans += 1
    return ans

inp = open("../input/07").read().split('\n')
print("Part 1:", part1(inp))
print("Part 2:", part2(inp))

File: 2016/rust/07.rs
// Advent of Code 2016, day 7
// (c) aichingert

fn has_abba(cur: &Vec<char>) -> bool {
    for i in 0..cur.len()-3 {
        if cur[i] != cur[i+1] && cur[i] == cur[i+3] && cur[i+1] == cur[i+2] { return true; }
    }
    false
}

fn part1(inp: &Vec<(Vec<String>, Vec<String>)>) -> u32 {
    let mut ans = 0;

    'outer: for l in inp {
        let mut v = false;

        for hypr in l.1.iter() {
            if has_abba(&hypr.chars().collect::<Vec<char>>()) { continue 'outer; }
        }

        for addr in l.0.iter() {
            if has_abba(&addr.chars().collect::<Vec<char>>()) {
                v = true;
                break;
            }
        }

        if v { ans += 1; }
    }

    ans
}

fn has_ssl(addr: &Vec<String>, hypr: &Vec<String>) -> bool {
    for i in 0..addr.len() {
        let c = addr[i].chars().collect::<Vec<char>>();

        for j in 0..c.len()-2 {
            if c[j] != c[j+1] && c[j] == c[j+2] {
                for hyp in hypr.iter() {
                    let ch = hyp.chars().collect::<Vec<char>>();

                    for k in 0..ch.len()-2 {
                        if ch[k] == ch[k+2] && ch[k] == c[j+1] && ch[k+1] == c[j] {
                            return true;
                        }
                    }
                }
            }
        }
    }
    false
}

fn part2(inp: &Vec<(Vec<String>, Vec<String>)>) -> usize {
    inp.iter().filter(|l| has_ssl(&l.0, &l.1)).count()
}

fn parse() -> Vec<(Vec<String>, Vec<String>)> {
    let mut ips = Vec::new();
    let inp = std::fs::read_to_string("../input/07").unwrap();

    for l in inp.lines() {
        let (lhs, rest): (&str, &str) = l.split_once('[').unwrap();
        let loc = ips.len();
        ips.push((vec![lhs.to_string()], vec![]));

        for split in rest.split('[') {
            let (hypr, addr): (&str, &str) = split.split_once(']').unwrap();
            ips[loc].0.push(addr.to_string());
            ips[loc].1.push(hypr.to_string());
        }

    }

    ips
}

fn main() {
    let inp = parse();

    println!("Part 1: {}", part1(&inp));
    println!("Part 2: {}", part2(&inp));
}
